### ISSUE TITLE
Training issue

### ISSUE THREAD

**Issue by danperazzo** [MichalOp/MineRL2020#1]
Hello Michael, thank you for releasing this wonderful code! 
I am participating this year in the MineRL competition and have been testing your code. But something strange happened.
I tried training your technique (changed the batch size from 4 to 3 due to memory constraints), and I only got rewards of 0.8. Did I do something wrong?

**Comment by piaomiaohuanchen** [comment]
Hello, I did not change the code, but the model performance is still very poor after the training is complete. I encountered a similar problem, how long will you train for?

**Comment by MichalOp** [comment]
Hi, I _think_ the code should work pretty much out of the box, assuming the data/some dependency has not changed significantly since I was training it. Just letting the unmodified `train.py` finish should be enough. I am not sure about reducing batch size, but I also wouldn't expect it to break things.

I started a training run from a completely clean environment - I reinstalled newest versions of all dependencies and re-downloaded the data, and one thing that I noticed is that with the newest MineRL package is not enough to run `minerl.data.download('data')`, as it only downloads 2 runs per category, to get all training data for my model I also had to run:
```python
minerl.data.download("data", environment="MineRLTreechopVectorObf-v0")
minerl.data.download("data", environment="MineRLObtainIronPickaxeVectorObf-v0")
minerl.data.download("data", environment="MineRLObtainDiamondVectorObf-v0")
```
so be sure to check that in your data directory you have over 100 replays in each of those sub-directories (MineRLTreechopVectorObf-v0, MineRLObtainIronPickaxeVectorObf-v0, MineRLObtainDiamondVectorObf-v0), and if not, run the code above. Also if you didn't download all the data be sure to delete the entire `train` directory from the previous run before retrying, as it will contain incorrectly computed KMeans.

I will provide an update once the run finishes and I will be able to check the performance.

**Comment by danperazzo** [comment]
Thank you @MichalOp! I have noticed that it only has 2 runs on my computer too. I will download the packages and restart the training. To get a sense, what is the hardware you are using (and used during the competition), and how long did it take to train?

**Comment by shwang** [comment]
In minerl v0.4.0, partly due to hosting costs, we changed the default target of `download` to be a minimal dataset, rather than the full dataset:
https://minerl.readthedocs.io/en/latest/tutorials/data_sampling.html#downloading-the-minerl-dataset-with-minerl-data-download

Following @MichalOp's comment above is one way to download the full dataset for the relevant environments. You can also use a command line interface as documented the in ReadTheDocs link. Apologies for the confusion this caused.

**Comment by danperazzo** [comment]
I got another issue: if I let it train for some time, it just appears: Killed, and the application stops.

**Comment by MichalOp** [comment]
> I will provide an update once the run finishes and I will be able to check the performance.

My run has completed, and it has successfully produced a model that runs reasonably well (although it got score of only 6 in the one test run I did). Training unfortunately is not deterministic, so it can be quite unreliable.

> To get a sense, what is the hardware you are using (and used during the competition), and how long did it take to train?

I used i9-9900K, 64 GB RAM, GTX 1080Ti and it took around 16 hours or so.

> I got another issue: if I let it train for some time, it just appears: Killed, and the application stops.

This sounds like OS killing it for some reason. Can you share like 10 output lines before the 'Killed' message so I can see at what stage it happens?

I would monitor how much memory it is using - it definitely can use more than 16GB of RAM. If this is the problem, you can try to reduce number of 'envs' in `BatchSeqLoader` (train.py, line 162) from 16 to let's say 8. On what machine are you trying to run it?

**Comment by danperazzo** [comment]
I was trying to run it on an i7, 16 Gb RAM, GTX 1050Ti (with 4 Gb of GPU memory). I will try to diminish my number of envs.

**Comment by piaomiaohuanchen** [comment]
> > I will provide an update once the run finishes and I will be able to check the performance.
> 
> My run has completed, and it has successfully produced a model that runs reasonably well (although it got score of only 6 in the one test run I did). Training unfortunately is not deterministic, so it can be quite unreliable.
> 
> > To get a sense, what is the hardware you are using (and used during the competition), and how long did it take to train?
> 
> I used i9-9900K, 64 GB RAM, GTX 1080Ti and it took around 16 hours or so.
> 
> > I got another issue: if I let it train for some time, it just appears: Killed, and the application stops.
> 
> This sounds like OS killing it for some reason. Can you share like 10 output lines before the 'Killed' message so I can see at what stage it happens?
> 
> I would monitor how much memory it is using - it definitely can use more than 16GB of RAM. If this is the problem, you can try to reduce number of 'envs' in `BatchSeqLoader` (train.py, line 162) from 16 to let's say 8. On what machine are you trying to run it?

Hello,  Can you perform more tests to see the performance of the trainde model.  I used all the data to train twice,  it took about 15-16 hours on a 1080Ti, but the final model did not work well. In the case of 100 tests, the average score was 2-3. Only slightly higher than the baseline.  Thank you.

**Comment by MichalOp** [comment]
> Can you perform more tests to see the performance of the trained model

I will do another test later in the week, the score of 2-3 is definitely way too low - I will check if I changed something by accident during cleanup after competition. 

Just to confirm, are you sure that you have deleted the `train` directory before retraining?

**Comment by piaomiaohuanchen** [comment]
> > Can you perform more tests to see the performance of the trained model
> 
> I will do another test later in the week, the score of 2-3 is definitely way too low - I will check if I changed something by accident during cleanup after competition.
> 
> Just to confirm, are you sure that you have deleted the `train` directory before retraining?

Yes, and I used all the training data from the beginning.

**Comment by MichalOp** [comment]
I did another train and test run, and I got score 7.2 out of it - still lower than my best score in the competition or in my earlier testing. I will do another one, but I also need my PC for other project that I am currently working on. 

It is possible that I simply got lucky in my local tests and in evaluation, or that I am now getting unlucky, because I wasn't able to spot any effective difference between the code here and the last code used in the competition. Non-deterministic loading was definitely a bad idea and I probably underestimated the impact it might have on training.

But scores 2-3 are really low, especially if you did two tries. Could you share output that the train.py prints at the end of the training (specifically including some of the lines that contain 3 numbers, first is loss, second is number of observations processed, third is number of observations per second), so that I can check the number of iterations done and the loss you are getting in training?

**Comment by piaomiaohuanchen** [comment]
Hello, since I didn’t save the training file before, I trained a model again, and after a hundred tests, the effect is still 2~3
This is the last few lines of my training input 

30.76780155301094 104992000 1649.9245719838252
2021-08-11 17:23:01 hzailab-tf21-pool.i.nease.net minerl.data.data_pipeline[4667] DEBUG Loading from file /storage06/users/yaoshunyi/data/MineRLObtainIronPickaxeVectorObf-v0/v3_remorseful_current_savage-1_12687-23101
2021-08-11 17:23:01 hzailab-tf21-pool.i.nease.net minerl.data.data_pipeline[4667] DEBUG Loading from file /storage06/users/yaoshunyi/data/MineRLObtainIronPickaxeVectorObf-v0/v3_juvenile_apple_angel-25_710393-715060
ok
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down

But at the same time I tried to change the kmeans n_cluser parameter 120->150.
Refer to this https://github.com/GJuceviciute/MineRL-2020
After retraining, the effect has a certain improvement, score9.5 after100 test .
thank you very much for your help

**Comment by piaomiaohuanchen** [comment]
Hello, since I didn’t save the training file before, I trained a model again, and after a hundred tests, the effect is still 2~3
This is the last few lines of my training output 

30.76780155301094 104992000 1649.9245719838252
2021-08-11 17:23:01 hzailab-tf21-pool.i.nease.net minerl.data.data_pipeline[4667] DEBUG Loading from file /storage06/users/yaoshunyi/data/MineRLObtainIronPickaxeVectorObf-v0/v3_remorseful_current_savage-1_12687-23101
2021-08-11 17:23:01 hzailab-tf21-pool.i.nease.net minerl.data.data_pipeline[4667] DEBUG Loading from file /storage06/users/yaoshunyi/data/MineRLObtainIronPickaxeVectorObf-v0/v3_juvenile_apple_angel-25_710393-715060
ok
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down
Shutting down

But at the same time I tried to change the kmeans n_cluser parameter 120->150.
Refer to this https://github.com/GJuceviciute/MineRL-2020
After retraining, the effect has a certain improvement, score9.5 after100 test .
thank you very much for your help
